### ISSUE TITLE
Add app/window/screen events

### ISSUE THREAD

**Issue by sdegutis** [sdegutis/mjolnir#72]
App events:
- [ ] AppLaunched
- [ ] AppTerminated
- [ ] AppHidden
- [ ] AppShown

Window events:
- [ ] WindowCreated
- [ ] WindowClosed
- [ ] WindowMoved
- [ ] WindowResized
- [ ] WindowMinimized
- [ ] WindowUnminimized
- [ ] WindowFocused

Screen events:
- [ ] ScreenLayoutChanged (i.e. screen may have been added/removed)

**Comment by sdegutis** [comment]
Regarding app lifetimes, it may be easier to use `kEventAppLaunched` and `kEventAppTerminated` than the way Zephyros does this.

**Comment by sdegutis** [comment]
Changing the milestone for this issue, there's no good reason to have it in 1.0 just yet.

**Comment by sdegutis** [comment]
The difficulty here is in order to observe window-related events we must first register a permanent observer (`AXObserverCreate`) for the app (`pid_t`) whose windows these may be; there's no way to register for a such an event independent of the application it may originate from. This involves a lot of ugly bookkeeping that I'd like to avoid putting into the core of the app if at all possible. But the only alternative solution seems to be asking the user to keep track of all current and future apps and maintain observers for them throughout their lifetimes, which also isn't reasonable.

**Comment by sdegutis** [comment]
If anyone has an alternative API to point me to, or any other ideas, your input will be greatly appreciated.

**Comment by sdegutis** [comment]
Solved.

**Comment by sdegutis** [comment]
@mgee It's coming soon.

**Comment by sdegutis** [comment]
This feature depends on #254.

**Comment by maxrothman** [comment]
@sdegutis :+1: as well, also can't find mjolnir-io/ext.

**Comment by sdegutis** [comment]
I'm working on events in a new app. Like everything else, doing events was way easier in Swift.

**Comment by sdegutis** [comment]
I'm currently working on events.

**Comment by nathankot** [comment]
@sdegutis thanks for the update :smile:

**Comment by svalaskevicius** [comment]
awesome! looking forward for event driven tiling WM on osx :)

**Comment by nathankot** [comment]
@svalaskevicius I hope we convert you eventually :stuck_out_tongue:

**Comment by svalaskevicius** [comment]
@nathankot thanks :) 
just to check - is the current implementation of tiling plugin working already or is it blocked by this issue.. and is there maybe a screencast of it in action? I've tried looking for it but couldn't find any.
(should I have asked this elsewhere?)

**Comment by nathankot** [comment]
@svalaskevicius yep it works, but doesn't respond to new windows being added (so you'll have to re-tile manually), the tiling project is young so don't think there are videos out in the wild but could do one of my workflow if you want?

**Comment by svalaskevicius** [comment]
it would be definitely interesting to see! - it's usually one of the first things I look for when I hear of a new tiling WM :)

**Comment by creese** [comment]
@nathankot The last time I checked, the tiling project did not work for multiple monitors. Has that issue been addressed?

**Comment by nathankot** [comment]
@creese don't use multiple screens myself but this may have fixed it: https://github.com/nathankot/mjolnir.tiling/issues/9 , if not , please make a new issue at mjolnir.tiling ;)

**Comment by simichaels** [comment]
Any progress on event handling so far?

**Comment by sdegutis** [comment]
They're basically finished in the Swift window manager I was writing. Main problem is there's no UI in that one yet.

**Comment by szymonkaliski** [comment]
Hmmm, but is everything else working? I for one don't need a UI to use it... Is it mjolnir compatible? I'd love to use some modules we already have...

**Comment by sdegutis** [comment]
@szymonkaliski It's all working, but I haven't finished porting all the functions (it's so tedious). It's probably already Mjolnir-compatible, but if not, it could easily be made to be so. I have two versions: one that uses Lua very similarly to Hammerspoon (a Mjolnir fork), and one that's meant to be one you write in pure Swift (in Xcode), and compile into an app that you can run. Since the work was split in two repos, each repo might have an API function or two that the other doesn't, so I'm not entirely which has what. Anyway, I plan to spend an hour or so today cleaning that up and making both repos presentable, and writing down the current status of each in their own READMEs. There'll likely still be some work needed to port some API functions over from Mjolnir, and if that's the case, I'd love some help.

**Comment by sdegutis** [comment]
Oh right, I remember what I planned for the Swift-only version: it would be a framework that you could just import, and you'd just make a new Swift app that embeds and links to this framework, and uses it in `applicationDidFinishLaunching`. That way, everyone could re-contribute to the framework itself, while keeping their "config" (i.e. the program putting the framework to good use) separate.

**Comment by sdegutis** [comment]
@szymonkaliski turned it into a framework at https://github.com/sdegutis/WindowManager fwiw

**Comment by cmsj** [comment]
FWIW, I expect that the Hammerspoon modules for app/windowscreen events would port over pretty easily. The code is at:

app events: https://github.com/Hammerspoon/hammerspoon/blob/master/extensions/application/watcher.m
window events: https://github.com/Hammerspoon/hammerspoon/tree/master/extensions/uielement
screen events: https://github.com/Hammerspoon/hammerspoon/blob/master/extensions/screen/watcher.m

(some of this stuff was inherited from Mjolnir's various modules anyway, so it may not all need to be ported over :)

**Comment by szymonkaliski** [comment]
@sdegutis https://github.com/sdegutis/WindowManager looks very cool, I hope I'll have some time to play with it soon!

**Comment by joehanchoi** [comment]
Any progress on this?
